docs/RingBuffer.md

```diff
@@ -0,0 +1,303 @@
+# New ebpf Ring Buffer Map (proposal)
+
+## Overview
+
+The current ringbuffer uses a pure callback-based approach to reading the ringbuffer.
+Linux also supports memory-mapped polling consumers, which can't be directly supported in the current model.
+
+The new API will support 2 consumer types: callbacks and direct access to the mapped producer memory (with poll to wait for data).
+
+Callback consumer:
+
+1. Call `ring_buffer__new` to set up callback
+2. Call `ring_buffer__consume` as needed to invoke the callback on any outsanding data that is ready
+
+Mapped memory consumer:
+
+1. Call `ring_buffer__new` to get a ringbuffer manager.
+2. Call `ebpf_ring_buffer_get_buffer` to get pointers to the mapped producer/consumer pages.
+3. Call `ebpf_ring_buffer_get_wait_handle` to get the wait handle.
+4. Directly read records from the producer pages (and update consumer offset as we read).
+5. Call `WaitForSingleObject`/`WaitForMultipleObject` as needed to wait for new data to be available.
+
+## API Changes
+
+### Changes to ebpf helper functions
+
+```c
+/**
+ * @brief Output record to ringbuf
+ *
+ * Note newly added flag values (to specify wakeup options).
+ *
+ * Wakeup options (flags):
+ * - 0 (auto/default): Notify if consumer has caught up.
+ * - BPF_RB_FORCE_WAKEUP - Always notify consumer.
+ * - BPF_RB_NO_WAKEUP - Never notify consumer.
+ *
+ */
+ebpf_result_t
+ebpf_ring_buffer_output(ebpf_ring_buffer_t* ring, uint8_t* data, size_t length, size_t flags)
+```
+
+### Updated supported libbpf functions
+
+```c
+
+struct ring_buffer;
+
+typedef int (*ring_buffer_sample_fn)(void *ctx, void *data, size_t size);
+
+struct ring_buffer_opts {
+	size_t sz; /* size of this struct, for forward/backward compatiblity */
+};
+
+#define ring_buffer_opts__last_field sz
+
+/**
+ * @brief Creates a new ring buffer manager.
+ *
+ * @param[in] map_fd File descriptor to ring buffer map.
+ * @param[in] sample_cb Pointer to ring buffer notification callback function.
+ * @param[in] ctx Pointer to sample_cb callback function context.
+ * @param[in] opts Ring buffer options.
+ *
+ * @returns Pointer to ring buffer manager.
+ */
+struct ring_buffer *
+ring_buffer__new(int map_fd, ring_buffer_sample_fn sample_cb, void *ctx,
```

Comment:
How do we differentiate between callback vs poll?

Reply:
Updated the design so callback consumers now use the libbpf ring buffer manager, and direct mapped memory consumers use the new functions.

Reply:
I don't follow.  How does Linux differentiate?  Why would we want a different answer?

Reply:
Linux doesn't have asynchronous callbacks, so they don't need differentiation. After you create the ringbuffer manager, then you call poll/consume to have the callback _synchronously_ called on available records.

Windows started with async callbacks, so this proposes a new option flag to keep the current default behavior and also support synchronous callbacks